core/trino-main/src/main/java/io/trino/sql/planner/rowpattern/IrRowPatternToProgramRewriter.java

```diff
@@ -0,0 +1,278 @@
+        new IrRowPatternToProgramRewriter(builder, labelMapping).process(node);
+        builder.add(new Done());
+        return new Program(builder.build());
+    }
+
+    private final Map<IrLabel, Integer> labelMapping;
+    private final ImmutableList.Builder<Instruction> instructions;
+    private int size;
+
+    IrRowPatternToProgramRewriter(ImmutableList.Builder<Instruction> instructions, Map<IrLabel, Integer> labelMapping)
+    {
+        this.instructions = requireNonNull(instructions, "instructions is null");
+        this.labelMapping = requireNonNull(labelMapping, "labelMapping is null");
+    }
+
+    @Override
+    protected Integer visitIrRowPattern(IrRowPattern node, Void context)
+    {
+        throw new IllegalStateException(format("unsupported node type: %s. run IrRowPatternFlattener and IrPatternAlternationOptimizer first", node.getClass().getName()));
+    }
+
+    @Override
+    protected Integer visitIrLabel(IrLabel node, Void context)
+    {
+        instructions.add(new MatchLabel(labelMapping.get(node)));
+        size++;
+        return size;
+    }
+
+    @Override
+    protected Integer visitIrEmptyPattern(IrEmptyPattern node, Void context)
+    {
+        return size;
+    }
+
+    @Override
+    protected Integer visitIrAnchorPattern(IrAnchorPattern node, Void context)
+    {
+        switch (node.getType()) {
+            case PARTITION_START:
+                instructions.add(new MatchStart());
+                break;
+            case PARTITION_END:
+                instructions.add(new MatchEnd());
+                break;
+            default:
+                throw new IllegalStateException("unexpected anchor type: " + node.getType());
+        }
+        size++;
+        return size;
+    }
+
+    @Override
+    protected Integer visitIrExcludedPattern(IrExcludedPattern node, Void context)
+    {
+        instructions.add(new Save());
+        size++;
+
+        size = process(node.getPattern());
+
+        instructions.add(new Save());
+        size++;
+
+        return size;
+    }
+
+    @Override
+    protected Integer visitIrPatternAlternation(IrPatternAlternation node, Void context)
+    {
+        if (node.getPatterns().size() == 1) {
+            return process(node.getPatterns().get(0));
+        }
+
+        Split split = new Split(size + 1, -1);
+        instructions.add(split);
+        size++;
+
+        process(node.getPatterns().get(0));
+        Jump jump = new Jump(-1);
+        instructions.add(jump);
+        size++;
+        split.setSecond(size);
+
+        int end = process(new IrPatternAlternation(node.getPatterns().subList(1, node.getPatterns().size())));
+        jump.setTarget(end);
+        return size;
+    }
+
+    @Override
+    protected Integer visitIrPatternConcatenation(IrPatternConcatenation node, Void context)
+    {
+        for (IrRowPattern pattern : node.getPatterns()) {
+            process(pattern);
+        }
+        return size;
+    }
```

Comment:
The use of `size` in this visitor is confusing and error-prone. In some places, it's modified inline. In others, it's returned and the caller assigns it.